lib/moab/storage_object_validator.rb

```diff
@@ -0,0 +1,196 @@
+      VERSIONS_NOT_IN_ORDER => "Should contain only sequential version directories. Current directories: %{addl}"
+    }.freeze
+
+    attr_reader :storage_obj_path
+
+    def initialize(storage_object)
+      @storage_obj_path = storage_object.object_pathname
+      @directory_entries_hash = {}
+    end
+
+    def validate_object
+      results = []
+      results.concat check_for_only_sequential_version_dirs
+      # if we only have sequential version directories uder the root object path, proceed
+      # to check for expected version subdirs, and to check contents of the data dir
+
+      results.concat correctly_formed_moab if results.empty?
+
+      results
+    end
+    # TODO: Figure out which methods should be public
+
+    private
+
+    def check_for_only_sequential_version_dirs
+      results = []
+
+      sub_dirs(storage_obj_path).each_with_index do |dir_name, index|
+        expected_vers_num = index + 1 # version numbering starts at 1, array indexing starts at 0
+        begin
+          if dir_name[1..-1].to_i != expected_vers_num
+            results << result_hash(VERSIONS_NOT_IN_ORDER, sub_dirs(storage_obj_path))
+            break
+          end
+        rescue ArgumentError
+          results << result_hash(VERSIONS_NOT_IN_ORDER, sub_dirs(storage_obj_path))
+        end
+      end
+
+      unless files_in_dir(storage_obj_path).empty?
+        results << result_hash(FILES_IN_VERSION_DIR, files_in_dir(storage_obj_path))
+      end
+
+      results
+    end
+
+    def correctly_formed_moab
+      results = []
+
+      version_directories = sub_dirs(storage_obj_path)
+      version_directories.each do |version_dir|
+        version_path = "#{storage_obj_path}/#{version_dir}"
+        version_sub_dirs = sub_dirs(version_path)
+        # Need to add before_result_size and after_result_size because if everything is expected
+        # and good, then no error message will be added to the result array, so if it stays the same
+        # size then it is ready to be checked for the data_sub_dirs and manfist_xml
+        before_result_size = results.size
+        results.concat check_sub_dirs(version_sub_dirs, version_dir, EXPECTED_VERSION_SUB_DIRS)
+        after_result_size = results.size
+        # don't bother checking the data sub dir if we didn't find the expected two subdirs
+
+        # Before it was result.size === 0, however for each version that is not correctly constructed,
+        # error messages will be concat to the array. By the time the loop reaches a version that is
+        # correctly constructe the size will stay the same as before (because it has the old error messages)
+        # so we must make sure before and after result size are the same - aka no error message was added
+        # and the version directory was correctly constructed
+
+        if before_result_size == after_result_size
+          data_dir_path = "#{version_path}/#{version_sub_dirs[0]}"
```

Comment:
I would rather you had a DATA_DIR constant and used it here:
```
DATA_DIR = 'data'.freeze # up high
data_dir_path = "#{version_path}/#{DATA_DIR}"
```

OR

```
data_dir_path = #{version_path}/#{EXPECTED_VERSION_SUB_DIRS[0]}
```
because in the suggestions above, it is clearer you getting the name from the expected constant.

using a position seems a bit more indirect that we'd desire.